build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1843,7 +1843,10 @@ message CacheCapabilities {
 
 // Capabilities of the remote execution system.
 message ExecutionCapabilities {
-  // Remote execution may only support a single digest function.
+  // Legacy field for indicating which digest function is supported by
+  // the remote execution system. Implementations should consider the
+  // repeated digest_functions field first, falling back to this
+  // singular field if unset.
   DigestFunction.Value digest_function = 1;
```

Comment:
Reading through https://developers.google.com/protocol-buffers/docs/encoding#optional it looks like it might actually be OK to upgrade a field to be repeated:

> Normally, an encoded message would never have more than one instance of a non-repeated field. However, parsers are expected to handle the case in which they do. For numeric types and strings, if the same field appears multiple times, the parser accepts the last value it sees.

Could we upgrade this field to repeated, and state that the preferred value for legacy clients should be specified last?

Reply:
I thought about this, but there's also a bit more risk to that. Back in #205 I made a very minor change to the protocol to pluralize the name of a repeated field. After we merged that, we got a report or two from people who were dependent on the Protobuf JSON encoding of these capabilities and ran into issues because of that change. My fear is that if we make an in-place change like that, we're going to run into the same kinds of issues again.

Happy to be persuaded otherwise, as I don't really depend on the JSON encoding of messages like these. Just want to make sure I don't make life of others harder.

Reply:
IIRC the problem in #205 was due to renaming the field, whereas we could keep the existing singular name and just make the field repeated (and state that servers should report the preferred value for legacy clients as the last one).

Reply:
I don't think you can if you want to keep compatibility, both for the binary and the JSON format. Here, consider this code in the Protobuf library:

https://github.com/protocolbuffers/protobuf-go/blob/master/encoding/protojson/decode.go#L506-L508

Whenever the descriptor states that a message field is repeated, it only allows the JSON representation to be a list. It doesn't fall back to interpreting it as a scalar. Conversely, `unmarshalScalar()` doesn't permit values to be lists.

Reply:
Yep, making a field repeated is a breaking change in the JSON serialisation - at least for most clients I'm aware of.

Reply:
Ah, ok. Scratch that idea then.